File: apps/android/app/src/main/java/com/smartble/core/utils/LogRedaction.kt
package com.smartble.core.utils

import org.json.JSONArray
import org.json.JSONObject
import org.json.JSONTokener

/**
 * F026 日志脱敏（BUSINESS_FLOW §7 / STORAGE_POLICY S-3 / R28）。
 *
 * uniapp `services/logger/log-redaction.js` 的 Kotlin 锁定镜像：
 * 敏感键（token/password/secret…）→ "***"，保护键白名单（deviceId/serviceUuid/sha256…）不脱敏。
 * Logger.emit 是唯一漏斗；任何旁路日志打印也必须先过这里。
 */
object LogRedaction {
    const val REDACT_PLACEHOLDER = "***"

    private val SENSITIVE_KEY_FRAGMENTS = listOf(
        "token", "password", "passwd", "secret", "apikey", "privatekey",
        "credential", "authorization", "cookie", "qrtoken",
    )

    private val PROTECTED_KEY_FRAGMENTS = listOf(
        "deviceid", "serviceuuid", "characteristicuuid", "serviceid",
        "characteristicid", "firmwareversion", "sha256",
    )

    private val STRING_SECRET_PATTERNS = listOf(
        Regex("""\bauthorization\s*:\s*Bearer\s+[A-Za-z0-9._-]+""", RegexOption.IGNORE_CASE),
        Regex("""\bBearer\s+[A-Za-z0-9._-]+""", RegexOption.IGNORE_CASE),
        Regex("""\b(access_token|refresh_token|qr_token|api_key|apikey|private_key|authorization|cookie)\s*[=:]\s*[^\s,&}"']+""", RegexOption.IGNORE_CASE),
        Regex("""\b(token|password|passwd|secret|credential|pw)\s*[=:]\s*[^\s,&}"']+""", RegexOption.IGNORE_CASE),
        Regex("\"(access_token|refresh_token|qr_token|api_key|apikey|private_key|authorization|cookie)\"\\s*:\\s*\"[^\"]*\"", RegexOption.IGNORE_CASE),
        Regex("\"(token|password|passwd|secret|credential)\"\\s*:\\s*\"[^\"]*\"", RegexOption.IGNORE_CASE),
    )

    private fun normalizeKey(key: Any?): String =
        (key?.toString() ?: "").lowercase().replace(Regex("[-_]"), "")

    fun isProtectedKey(key: Any?): Boolean {
        val normalized = normalizeKey(key)
        if (normalized.isEmpty()) return false
        if (PROTECTED_KEY_FRAGMENTS.contains(normalized)) return true
        if (normalized.endsWith("uuid") &&
            (normalized.contains("service") || normalized.contains("characteristic"))
        ) {
            return true
        }
        return false
    }

    fun isSensitiveKey(key: Any?): Boolean {
        if (isProtectedKey(key)) return false
        val normalized = normalizeKey(key)
        if (normalized.isEmpty()) return false
        return SENSITIVE_KEY_FRAGMENTS.any { normalized.contains(it) }
    }

    fun sanitizeLogValue(value: Any?, seen: MutableSet<Int> = mutableSetOf()): Any? = when (value) {
        null -> null
        is String -> sanitizeLogString(value)
        is Number, is Boolean -> value
        is Throwable -> mapOf(
            "name" to value.javaClass.simpleName,
            "message" to (sanitizeLogString(value.message ?: value.toString()) ?: ""),
        )
        is List<*> -> value.map { sanitizeLogValue(it, seen) }
        is Array<*> -> value.map { sanitizeLogValue(it, seen) }
        is Map<*, *> -> sanitizeLogMap(value, seen)
        else -> value
    }

    fun sanitizeLogMap(input: Map<*, *>, seen: MutableSet<Int> = mutableSetOf()): Any {
        val marker = System.identityHashCode(input)
        if (seen.contains(marker)) return "[Circular]"
        seen.add(marker)

        val output = LinkedHashMap<String, Any?>()
        for ((key, value) in input) {
            output[key.toString()] =
                if (isSensitiveKey(key)) REDACT_PLACEHOLDER else sanitizeLogValue(value, seen)
        }
        return output
    }

    fun sanitizeLogString(input: String?): String? {
        val text = input ?: return null
        if (text.isEmpty()) return text

        val trimmed = text.trim()
        if ((trimmed.startsWith("{") && trimmed.endsWith("}")) ||
            (trimmed.startsWith("[") && trimmed.endsWith("]"))
        ) {
            try {
                return sanitizeJsonValue(JSONTokener(trimmed).nextValue()).toString()
            } catch (_: Exception) {
                // fall through to pattern redaction
            }
        }

        var redacted = text
        for (pattern in STRING_SECRET_PATTERNS) {
            redacted = pattern.replace(redacted) { m ->
                val matched = m.value
                when {
                    Regex("""^Bearer\s+""", RegexOption.IGNORE_CASE).containsMatchIn(matched) ->
                        "Bearer $REDACT_PLACEHOLDER"
                    Regex("""^authorization\s*:\s*Bearer""", RegexOption.IGNORE_CASE).containsMatchIn(matched) ->
                        "authorization: Bearer $REDACT_PLACEHOLDER"
                    else -> {
                        val label = matched.split(Regex("[=:]"))[0].trim()
                        val sep = if (matched.contains(":") && !matched.contains("=")) ":" else "="
                        "$label$sep$REDACT_PLACEHOLDER"
                    }
                }
            }
        }
        return redacted
    }

    private fun sanitizeJsonValue(value: Any?): Any = when (value) {
        is JSONObject -> {
            val out = JSONObject()
            for (key in value.keys()) {
                if (isSensitiveKey(key)) out.put(key, REDACT_PLACEHOLDER)
                else out.put(key, sanitizeJsonValue(value.get(key)))
            }
            out
        }
        is JSONArray -> {
            val out = JSONArray()
            for (i in 0 until value.length()) out.put(sanitizeJsonValue(value.get(i)))
            out
        }
        null, JSONObject.NULL -> JSONObject.NULL
        else -> value
    }
}

File: apps/flutter/lib/core/utils/log_redaction.dart
/// F026 日志脱敏（BUSINESS_FLOW §7 / STORAGE_POLICY S-3 / R28）。
///
/// uniapp `services/logger/log-redaction.js` 的 Dart 锁定镜像：
/// 敏感键（token/password/secret…）→ `***`，保护键白名单（deviceId/serviceUuid/sha256…）不脱敏。
/// 任意日志输出路径（Logger 漏斗、payload/异常旁路打印）都必须先过这里。
library;

import 'dart:convert';

const String redactPlaceholder = '***';

const List<String> _sensitiveKeyFragments = [
  'token',
  'password',
  'passwd',
  'secret',
  'apikey',
  'privatekey',
  'credential',
  'authorization',
  'cookie',
  'qrtoken',
];

const List<String> _protectedKeyFragments = [
  'deviceid',
  'serviceuuid',
  'characteristicuuid',
  'serviceid',
  'characteristicid',
  'firmwareversion',
  'sha256',
];

final List<RegExp> _stringSecretPatterns = [
  RegExp(r'\bauthorization\s*:\s*Bearer\s+[A-Za-z0-9._-]+', caseSensitive: false),
  RegExp(r'\bBearer\s+[A-Za-z0-9._-]+', caseSensitive: false),
  RegExp(r'''\b(access_token|refresh_token|qr_token|api_key|apikey|private_key|authorization|cookie)\s*[=:]\s*[^\s,&}"']+''', caseSensitive: false),
  RegExp(r'''\b(token|password|passwd|secret|credential|pw)\s*[=:]\s*[^\s,&}"']+''', caseSensitive: false),
  RegExp(r'"(access_token|refresh_token|qr_token|api_key|apikey|private_key|authorization|cookie)"\s*:\s*"[^"]*"', caseSensitive: false),
  RegExp(r'"(token|password|passwd|secret|credential)"\s*:\s*"[^"]*"', caseSensitive: false),
];

String _normalizeKey(Object? key) {
  return (key?.toString() ?? '').toLowerCase().replaceAll(RegExp('[-_]'), '');
}

bool _isProtectedKey(Object? key) {
  final normalized = _normalizeKey(key);
  if (normalized.isEmpty) return false;
  if (_protectedKeyFragments.contains(normalized)) return true;
  if (normalized.endsWith('uuid') &&
      (normalized.contains('service') || normalized.contains('characteristic'))) {
    return true;
  }
  return false;
}

bool isSensitiveKey(Object? key) {
  if (_isProtectedKey(key)) return false;
  final normalized = _normalizeKey(key);
  if (normalized.isEmpty) return false;
  return _sensitiveKeyFragments.any(normalized.contains);
}

dynamic sanitizeLogValue(dynamic value, [Set<int>? seen]) {
  if (value == null) return value;
  if (value is String) return sanitizeLogString(value);
  if (value is num || value is bool) return value;
  if (value is Error) {
    return <String, dynamic>{
      'name': value.runtimeType.toString(),
      'message': sanitizeLogString(value.toString()),
    };
  }
  if (value is List) {
    return value.map((item) => sanitizeLogValue(item, seen)).toList();
  }
  if (value is Map) {
    return _sanitizeMap(value, seen ?? <int>{});
  }
  return value;
}

Map<String, dynamic> sanitizeLogObject(Map? input, [Set<int>? seen]) {
  if (input == null) return <String, dynamic>{};
  return _sanitizeMap(input, seen ?? <int>{});
}

dynamic _sanitizeMap(Map input, Set<int> seen) {
  final marker = identityHashCode(input);
  if (seen.contains(marker)) return '[Circular]';
  seen.add(marker);

  final output = <String, dynamic>{};
  input.forEach((key, value) {
    if (isSensitiveKey(key)) {
      output['$key'] = redactPlaceholder;
    } else {
      output['$key'] = sanitizeLogValue(value, seen);
    }
  });
  return output;
}

String? sanitizeLogString(String? input) {
  if (input == null) return null;
  final text = input;
  if (text.isEmpty) return text;

  final trimmed = text.trim();
  if ((trimmed.startsWith('{') && trimmed.endsWith('}')) ||
      (trimmed.startsWith('[') && trimmed.endsWith(']'))) {
    try {
      final parsed = jsonDecode(trimmed);
      return jsonEncode(sanitizeLogValue(parsed));
    } catch (_) {
      // fall through to pattern redaction
    }
  }

  var redacted = text;
  for (final pattern in _stringSecretPatterns) {
    redacted = redacted.replaceAllMapped(pattern, (match) {
      final matched = match.group(0)!;
      if (RegExp(r'^Bearer\s+', caseSensitive: false).hasMatch(matched)) {
        return 'Bearer $redactPlaceholder';
      }
      if (RegExp(r'^authorization\s*:\s*Bearer', caseSensitive: false).hasMatch(matched)) {
        return 'authorization: Bearer $redactPlaceholder';
      }
      final label = matched.split(RegExp(r'[=:]'))[0].trim();
      final sep = matched.contains(':') && !matched.contains('=') ? ':' : '=';
      return '$label$sep$redactPlaceholder';
    });
  }
  return redacted;
}
